File: Data Structures & Algorithms/time-based-key-value-store/submission-0.cpp
class TimeMap {
private:
    unordered_map<string, vector<pair<int, string>>> keyStore;

public:
    TimeMap() {}

    void set(string key, string value, int timestamp) {
        keyStore[key].push_back({timestamp, value});
    }

    string get(string key, int timestamp) {
        auto& values = keyStore[key];
        if (values.empty()) return "";

        int left = 0;
        int right = values.size() - 1;
        string result = "";

        // simpler version of binary search
        while (left <= right) {
            int mid = (left + right) / 2;  // simpler, still works fine here
            if (values[mid].first == timestamp) {
                return values[mid].second; // exact match found
            }
            else if (values[mid].first < timestamp) {
                result = values[mid].second;
                left = mid + 1; // move right
            }
            else {
                right = mid - 1; // move left
            }
        }

        return result;
    }
};

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-0.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        left=0
        right=len(nums)-1
        while left<right:
            middle=(left+right)//2
            if (nums[left]<=nums[middle]<=nums[right]): #regular binary search
                right=middle-1
            else:
                if nums[left]>nums[middle]:
                    right=middle
                else:
                    left=middle+1
        return nums[left]